docs/proposals/ingress-l4.md

```diff
@@ -0,0 +1,346 @@
+| --------- | --- | --- | --- |
+| AWS       |  x  |     |  x  |
+| GCE       |  x  |  x  | (1) |
+| Mesos     |  ?  |  ?  |  ?  |
+| Openstack |  x  |  x  |  x  |
+| Rackspace |  x  |  x  |     |
+| vSphere   |  x  |  x  |  x  |
+
+(1) GCE has TLS as an alpha service
+
+### TCP and UDP support
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 53
+      backend:
+        serviceName: testdns
+        servicePort: 53
+    - udp:
+        port: 53
+      backend:
+        serviceName: testdns
+        servicePort: 53
+```
+
+While HTTP rules have host specifiers, in TCP and UDP rules they would not be
+allowed as hostname handling is specific to L5 or higher layer protocol being
+used over TCP or UDP.  TCP and UDP as protocol specifiers should be agnostic to
+higher layer protocols.
+
+### Ingress type specifier
+
+The current Ingress API assumes a protocol based on the presence or absence
+of a key (tls, http).  This leaves open the possibility of a spec that defines
+all protocols.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 3306
+      udp:
+        port: 3306
+      host: www.foo.com
+      tls:
+        - secretName: testsecret
+      backend:
+        serviceName: testmysql
+        servicePort: 3306
+      http:
+        paths:
+        - path: /bar
+          backend:
+              serviceName: bar
+              servicePort: 80
+```
+
+A specification like the one above is arguably too complex and should be
+written as separate rules for each protocol being supported.  However, the
+loose syntax of the specification allows construction of a complex
+specification.  Using a 'type' or 'protocol' specifier would remove this
+ambiguity in the specification.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - protocol: tcp
+      port: 3306
+      backend:
+        serviceName: testmysql
+        servicePort: 3306
+
+    - protocol: udp
+      port: 3306
+      backend:
+        serviceName: testmysql
+        servicePort: 3306
+
+    - protocol: tls
+      host: www.foo.com
+      tls:
+        - secretName: testsecret
+      http:
+        paths:
+        - path: /bar
+          backend:
+              serviceName: bar
+              servicePort: 80
+```
+
+### L4 TLS Termination
+
+While it would be very useful to be able to express TLS protected L4 Ingress,
+such that TLS can be terminated by the Ingress implementation, not all cloud
+providers support TLS termination, thus more design is required to add L4 TLS
+for Ingress in a manner that would work for all cloud providers.
+
+## Proposed Addition/Change
+
+ - Add a protocol specifier for rules to explicitly declare the desired protocol
+   for Ingress
+ - Add TCP and UDP protocol types to allow definition of TCP and UDP Ingress
```

Comment:
I agree about the reduction of the topic.  When I started this, I failed to appreciate how difficult it would be to get quick consensus on multiple issues.  I've started writing separate proposals for the other pieces.